<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="control_mode" default="0" />
  <arg name="type" default="quad" />
  <arg name="model" default="euclid201806" />
  <arg name="estimate_mode"  default= "0" />
  <arg name="headless" default="True" />
  <arg name="worldtype" default="$(find aerial_robot_simulation)/gazebo_model/world/empty.world" />
  <arg name="direct_model" default="False" />
  <arg name="direct_model_name" />


  ###########  UAV Config  ###########
  # HYDRUS = 16 # 0x10, hydrus type with  4dof underactuated property
  <param name="/uav_info/uav_model" value= "16" />

  ###########  Motor Config  ###########
  <rosparam file="$(find hydrus)/config/MotorInfo.yaml" command="load" />

  ###########  Battery Config  ###########
  <rosparam file="$(find hydrus)/config/$(arg type)/Battery.yaml" command="load" />

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" output="screen" >

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimator/estimate_mode" value= "2" if="$(arg simulation)" />
    <param name="estimator/estimate_mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>

    # World Pos Control Mode: 0
    # World Vel Control Mode: 2
    # Local Vel Control Mode: 3
    # Attitude Control Mode: 4
    <param name="navigator/xy_control_mode"  value="$(arg control_mode)"/>

    <param name="simulation" value="$(arg simulation)" />  <!--for sensor plugin-->
    <param name="param_verbose" value="false"/>
    <param name="main_rate" type="double" value="40"/>

    ###########  Sensor Fusion  ###########
    <rosparam file="$(find hydrus)/config/$(arg type)/egomotion_estimation/$(arg model).yaml" command="load" if="$(arg real_machine)"/>
    <rosparam file="$(find hydrus)/config/$(arg type)/egomotion_estimation/simulation.yaml" command="load" if="$(eval arg('simulation') * (1 - arg('real_machine')))"/>
    ###########  PID Control  ###########
    <rosparam file="$(find hydrus)/config/$(arg type)/DifferentialFlatnessPidControlConfig.yaml" command="load" />

    ###########  Teleop  ###########
    <rosparam file="$(find hydrus)/config/$(arg type)/TeleopNavigationConfig.yaml" command="load" />
  </node>

  ###########  Robot Model  ###########
  <include file="$(find hydrus)/launch/robot_model.launch" >
    <arg name="type" value="$(arg type)" />
    <arg name="model" value="$(arg model)"/>
    <arg name="headless" value="$(arg headless)" />
    <arg name="need_joint_state" value="false" if ="$(eval arg('simulation') + arg('real_machine') > 0)"/>
    <arg name="description_mode" value="gazebo" if="$(arg simulation)"/>
    <arg name="direct_model" value="$(arg direct_model)" />
    <arg name="direct_model_name" value="$(arg direct_model_name)" if="$(arg direct_model)"/>
  </include >

  ###########  Transform Control  ###########
  <node pkg="hydrus" type="transform_control_node" name="hydrusx"  output="screen" respawn="false">
    <rosparam file="$(find hydrus)/config/$(arg type)/Hydrus.yaml" command="load" />
    <param name="kinematic_verbose" value="true" />
    <param name="control_verbose" value="false" />
    <param name="debug_verbose" value="false" />
    <param name="verbose" value="true" />
  </node>

  ###########  Sensors  ###########
  <include file="$(find hydrus)/launch/includes/$(arg model)/sensors.launch.xml" >
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="hydrus" type="dynamixel_bridge_node" name="servo_bridge"  output="screen"  ns="hydrusx" if="$(arg real_machine)" >
    <rosparam file="$(find hydrus)/config/$(arg type)/Servo.yaml" command="load" />
  </node>

  ########## Simulation in Gazebo #########
  <rosparam file="$(find hydrus)/config/$(arg type)/SimulationControl.yaml" command="load"/>
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(eval arg('simulation') * (1 - arg('real_machine')))" >
    <arg name="model" default="hydrusx" />
    <arg name="gui" default="false" if="$(arg headless)" />
    <arg name="headless" default="$(arg headless)" />
    <arg name="worldtype" value="$(arg worldtype)" />
  </include>

</launch>
